### ISSUE TITLE
LTO mismatch with `dgemm` with RcppArmadillo and R core definitions

### ISSUE THREAD

**Issue by mattfidler** [RcppCore/RcppArmadillo#343]
Hi @eddelbuettel 

I think there may be a RcppArmadillo LTO issue that is described by the following log:

https://www.stats.ox.ac.uk/pub/bdr/LTO/RxODE.out

To save you time in parsing, it boils down to the following:

```sh 
/data/gannet/ripley/R/R-devel/include/R_ext/BLAS.h:207:1: warning: type of ‘dgemm_’ does not match original declaration [-Wlto-type-mismatch]
  207 | F77_NAME(dgemm)(const char *transa, const char *transb, const int *m,
      | ^
/data/gannet/ripley/R/test-4.2/RcppArmadillo/include/armadillo_bits/def_blas.hpp:115:8: note: type mismatch in parameter 14
  115 |   void arma_fortran(arma_dgemm)(const char* transA, const char* transB, const blas_int* m, const blas_int* n, const blas_int* k, const double*   alpha, const double*   A, const blas_int* ldA, const double*   B, const blas_int* ldB, const double*   beta, double*   C, const blas_int* ldC, blas_len transA_len, blas_len transB_len) ARMA_NOEXCEPT;
      |        ^
/data/gannet/ripley/R/test-4.2/RcppArmadillo/include/armadillo_bits/def_blas.hpp:115:8: note: type ‘blas_len’ should match type ‘void’
/data/gannet/ripley/R/test-4.2/RcppArmadillo/include/armadillo_bits/def_blas.hpp:115:8: note: ‘dgemm_’ was previously declared here
```

My guess is the LTO compile options have changed and become more restrictive (ie `flto=10`). I did (with your help) also change the headers to strict R headers, though I think that is unrelated

Perhaps my reading of the error message is incorrect.  I thought I would bring it to your attention.  

While I do not wish to, I may have to move and modify the armadillo headers (because of the fix by 2021-09-02  CRAN deadline).

**Comment by eddelbuettel** [comment]
Hm, I have yet to turn on LTO on the Debian (and Ubuntu) side of things, i.e. I have not enabled it for R, or Rcpp, or RcppArmadillo.  So maybe it is something triggered by Prof Ripley on his side.

That said, it may be reproducible, and @conradsnicta is generally very receptive to clear and concise examples. So could I ask you to maybe try to cook up a (non-R, standalone) example program that with/without LTO options shows the difference?  I probably have dgemm callers for BLAS here somewhere though we probably want to tickle it via RcppArmadillo.

**Comment by eddelbuettel** [comment]
Took another look at your [full report here](https://www.stats.ox.ac.uk/pub/bdr/LTO/RxODE.out) (reference from the standard results page [here](https://cran.r-project.org/web/checks/check_results_RxODE.html)) and its a delicate issue as R (and LTO) look back to the basic C linkage yet Armadillo has a different (stacked, via arma_fortran() calling arma_dgemm()) setup. I am sure that will come up again.  Hm.  I may need to consult with @conradsnicta and see what we can do here medium-term as LTO will surely become more common and standard...

**Comment by mattfidler** [comment]
For a quick fix you could inline the blas routines and rename them; I don't think it is an ideal solution, but it worked for some of the matrix exponential calls used in the RxODE ode solving  
                                                                               
https://github.com/nlmixrdevelopment/RxODE/blob/346edfe7be78c8998eccbb51b89a36929de84488/src/matexp.f#L12

This is not ideal it means you cant choose the blas that you want to do calculations.

When reading through the C code myself, I noticed the number of arguments for `dgemm` may be different in R than in standard blas.  I think they used this trick as  a safety check to get past the  problems passing strings between C and fortran that we saw in #254, but I'm unsure their motivation

**Comment by eddelbuettel** [comment]
Quick fixes sometimes bite your backside in the medium / longer-term.  A minimal yet reproducible example (outside of R) would be a good first step in getting towards this.

**Comment by mattfidler** [comment]
Hm.  I could likely do this inside R but I dont think I can outside R.  The blas R and standard blas definitions are clashing.  I will try a simple R example first.

**Comment by eddelbuettel** [comment]
An R-only example would still help greatly. I can then either try to work from that or discuss it with @conradsnicta .

**Comment by eddelbuettel** [comment]
The other thing to do (for at least, while we work through this) is that you could add a `-fno-lto` flag.

**Comment by mattfidler** [comment]
Well I found a work around for the package for now.  Interestingly enough, this came because of a code refactoring that actually didn't directly involve armadillo (found from a bisection)

https://github.com/nlmixrdevelopment/RxODE/commit/52294dcc98e45681ee361c0d1942ff165ece88f0

Alas however, I am unsure how this came up.  I believe there is some bug somewhere, but since I can't figure out how to create it in a small reproducible example, and I know how to fix it for now, feel free to close this bug (unless you want to try to figure it out)

**Comment by eddelbuettel** [comment]
Yeah happy to close if that is ok with you.

Sometime strange things happen just with header reordering and alike -- maybe something was shadowing something elsewhere. It does smell a little local to your repo which is why closing may be wisest.  I still have (optional?) conversion to lto for the R installation as a TODO so maybe we get to revisit this...  

And by all means let's reopen if there is something new to review.

**Comment by mattfidler** [comment]
Thanks Dirk

**Comment by conradsnicta** [comment]
@mattfidler @eddelbuettel 

TLDR: Not sure what's going on. On the surface the definitions of `dgemm` in both R and Armadillo are the same.  Possible differences:
* R has explicit `extern` while Armadilo doesn't.  However, the definition in Armadillo is within an `extern "C"` block
* R may be using a different definition of the argument type to handle the so-called Fortran hidden arguments (see https://gitlab.com/conradsnicta/armadillo-code/-/issues/123).  
The definition of the argument type for hidden arguments:
  * In R it is `FCLEN`. I don't know what `FC_LEN_T` is for sure without further digging. Quick search finds `size_t`, but I'm not certain on this.
  * in Armadillo it is `size_t` for GCC >= 8, and `int` for GCC <= 7.
  * As a potential workaround, you can disable the use of the so-called hidden arguments by defining `ARMA_DONT_USE_FORTRAN_HIDDEN_ARGs` before including Armadillo / RcppArmadillo.

Long explanation:

R has `dgemm` defined as (see https://github.com/wch/r-source/blob/R-4-1-branch/src/include/R_ext/BLAS.h):
```
BLAS_extern void
F77_NAME(dgemm)(const char *transa, const char *transb, const int *m,
		const int *n, const int *k, const double *alpha,
		const double *a, const int *lda,
		const double *b, const int *ldb,
		const double *beta, double *c, const int *ldc 
		FCLEN FCLEN);
```
where `FCLEN` is defined as:
```
#ifdef FC_LEN_T
# include <stddef.h> // for size_t if needed
# define FCLEN ,FC_LEN_T
# define FCONE ,(FC_LEN_T)1
#else
# define FCLEN
# define FCONE
#endif
```

**NOTE:** I don't know what `FC_LEN_T` is without further digging.

Armadillo has `dgemm` defined by default as (see https://gitlab.com/conradsnicta/armadillo-code/-/blob/10.6.x/include/armadillo_bits/def_blas.hpp):
```
void arma_fortran(arma_dgemm)(const char* transA, const char* transB, const blas_int* m,
		const blas_int* n, const blas_int* k, const double*   alpha,
		const double*   A, const blas_int* ldA,
		const double*   B, const blas_int* ldB,
		const double*   beta, double*   C, const blas_int* ldC,
		blas_len transA_len, blas_len transB_len)
		ARMA_NOEXCEPT;
```
where:
* `arma_dgemm` is defined as 
```
#define arma_dgemm dgemm
```
* `blas_int` is typedefed by default to `int` (see https://gitlab.com/conradsnicta/armadillo-code/-/blob/10.6.x/include/armadillo_bits/typedef_elem.hpp):
```
#if   defined(ARMA_BLAS_LONG_LONG)
  typedef long long blas_int;
  #define ARMA_MAX_BLAS_INT 0x7fffffffffffffffULL
#elif defined(ARMA_BLAS_LONG)
  typedef long      blas_int;
  #define ARMA_MAX_BLAS_INT 0x7fffffffffffffffUL
#else
  typedef int       blas_int;
  #define ARMA_MAX_BLAS_INT 0x7fffffffU
#endif
```
* `blas_len` is typedefed on recent GCC versions to `size_t`, but varies for earlier GCC versions (see https://gitlab.com/conradsnicta/armadillo-code/-/blob/10.6.x/include/armadillo_bits/typedef_elem.hpp):
```
#if !defined(ARMA_FORTRAN_CHARLEN_TYPE)
  #if defined(__GNUC__) && !defined(__clang__)
    #if (__GNUC__ <= 7)
      #define ARMA_FORTRAN_CHARLEN_TYPE int
    #else
      #define ARMA_FORTRAN_CHARLEN_TYPE size_t
    #endif
  #else
    // TODO: determine the type for other compilers
    #define ARMA_FORTRAN_CHARLEN_TYPE size_t
  #endif
#endif

typedef ARMA_FORTRAN_CHARLEN_TYPE blas_len;
```

* Lastly, `ARMA_NOEXCEPT` is empty by default:
```
#if defined(ARMA_BLAS_NOEXCEPT)
  #undef  ARMA_NOEXCEPT
  #define ARMA_NOEXCEPT noexcept
#else
  #undef  ARMA_NOEXCEPT
  #define ARMA_NOEXCEPT
#endif
```

**Comment by conradsnicta** [comment]
@mattfidler Looking closer at the error output, we have this:

> /data/gannet/ripley/R/test-4.2/RcppArmadillo/include/armadillo_bits/def_blas.hpp:115:8: note: type ‘blas_len’ should match type ‘void’

So it looks like `blas_len` is being re-defined, or typedefed, or otherwise mucked up somewhere.

`blas_len` is used by Armadillo in its definition of `dgemm`.  By default it is typedefed to `int`.

**Comment by conradsnicta** [comment]
@mattfidler Do you by any chance use complex matrices in your code?  The type `void*` is typically used to pass complex arrays from C to Fortran.

Maybe there is a mismatch/bug in your code between `dgemm` and `zgemm` ?

**Comment by conradsnicta** [comment]
@mattfidler Looking closer at the error output, we have this:

> /data/gannet/ripley/R/test-4.2/RcppArmadillo/include/armadillo_bits/def_blas.hpp:115:8: note: type ‘blas_len’ should match type ‘void’

`blas_len` is used by Armadillo in its definition of `dgemm` to handle the so-called hidden arguments. By default it is typedefed to `size_t` in recent GCC versions.

One possibility is that `blas_len` is being re-defined, or typedefed, or otherwise mucked up somewhere.

Another possibility is that the definition of `dgemm` in R somehow loses the so-called hidden arguments.  This can happen if `FC_LEN_T` is not defined, or is deliberately undefined: https://github.com/wch/r-source/blob/trunk/src/include/R_ext/BLAS.h

R appears to handle the hidden arguments via the `FC_LEN_T` type, and then uses `FCLEN` in its definition of `dgemm`.

**Comment by mattfidler** [comment]
You are right @conradsnicta,  

The one thing I don't understand is why it thinks it needs to be void.  To me, that means that is somehow being mucked up somewhere.  

That could definitely cause the LTO errors that were observed.

At the same time, it doesn't always occur and I'm unsure what is causing it here.
